opentracing-impl/src/test/java/io/opentracing/AbstractTracerTest.java

```diff
@@ -89,6 +89,23 @@ public void testNonEmptyExtract() {
         assertEquals("Should find the marker", "whatever", ((TestSpanBuilder)result).operationName);
     }
 
+    @Test
+    public void testExtractAsParent() throws Exception {
+        Map<String,String> map = Collections.singletonMap("test-marker", "whatever");
+        TextMapExtractAdapter adapter = new TextMapExtractAdapter(map);
+        AbstractTracer tracer = new TestTracerImpl();
+        SpanContext parent = tracer.extract(Format.Builtin.TEXT_MAP, adapter);
+        assert AbstractNoopSpan.INSTANCE != tracer.buildSpan("child").asChildOf(parent).start();
+    }
+
+    @Test
+    public void testExtractOfNoParent() throws Exception {
+        AbstractTracer tracer = new TestTracerImpl();
+        assert AbstractNoopSpan.INSTANCE == tracer.buildSpan("child").asChildOf((Span)AbstractNoopSpan.INSTANCE).start();
+        assert AbstractNoopSpan.INSTANCE == tracer.buildSpan("child").asChildOf((SpanContext)AbstractNoopSpan.INSTANCE).start();
+        assert AbstractNoopSpan.INSTANCE == tracer.buildSpan("child").asChildOf(AbstractNoopSpanBuilder.INSTANCE).start();
```

Comment:
I think you can achieve the same result by always writing instrumentation in form (A), where a NoopSpanContext is always passed to childOf(), which shields applications/instrumentation code from the need to even know about the existence of such thing as noopX (unless they really want to know, then they can use model C). Then the question of whether to start or not to start a new trace as a child of Noop context can be a configuration parameter of the tracer, addressed by service deployment, not changes in instrumentation (thus no code changes).

FWIW, at Uber all instrumentation is written to always start a new trace if none present in the inbound request. The reasoning here is that as a service owner I make a decision that I want tracing enabled in my service, regardless of who happened to call me, with or without upstream trace. Doing so allows various analytics of the traffic patterns because traces in a given service are predictable, not at the whim of the caller.

Anyway, agreed that this is not a blocker for this PR.